src/subdataframe/subdataframe.jl

```diff
@@ -72,26 +72,47 @@ function SubDataFrame(parent::DataFrame, rows::Colon)
     return SubDataFrame(parent, 1:nrow(parent))
 end
 
+function SubDataFrame(parent::DataFrame, row::Bool)
+    throw(ArgumentError("invalid row index: $row of type `Bool`"))
+end
+
 function SubDataFrame(parent::DataFrame, row::Integer)
+    Base.depwarn("Creation of `SubDataFrame` with a single row is deprecated. " *
+                 "In the future `DataFrameRow` will be used in this case.", :SubDataFrame)
     return SubDataFrame(parent, [Int(row)])
 end
 
 function SubDataFrame(parent::DataFrame, rows::AbstractVector{<:Integer})
+    if any(x -> x isa Bool, rows)
+        throw(ArgumentError("invalid row index of type `Bool`"))
+    end
     return SubDataFrame(parent, convert(Vector{Int}, rows))
 end
 
 function SubDataFrame(parent::DataFrame, rows::AbstractVector{Bool})
+    if length(rows) != nrow(parent)
+        throw(ArgumentError("invalid length of `AbstractVector{Bool}` row index" *
+                            " (got $(length(rows)), expected $(nrow(parent))"))
+    end
     return SubDataFrame(parent, findall(rows))
 end
 
-function SubDataFrame(sdf::SubDataFrame, rowinds::Union{T, AbstractVector{T}}) where {T <: Integer}
-    return SubDataFrame(parent(sdf), rows(sdf)[rowinds])
+function SubDataFrame(parent::DataFrame, rows::AbstractVector)
+    if !all(x -> (x isa Integer) && !(x isa Bool), rows)
+        throw(ArgumentError("only `Integer` indices are accepted in `rows`"))
+    end
+    return SubDataFrame(parent, convert(Vector{Int}, rows))
 end
 
-function SubDataFrame(sdf::SubDataFrame, rowinds::Colon)
-    return sdf
+SubDataFrame(parent::DataFrame, rows) =
+    SubDataFrame(parent, (1:nrow(parent))[rows])
```

Comment:
What's this trick?

Reply:
The idea is that:
* we are sure that we get `Vector{Int}` as a result;
* `(1:nrow(parent))[rows]` will use machinery from Base to catch any problematic values in `rows`, OTOH anything that is valid in Base will be valid here.

If not for performance and error clarity reasons this could be the only definition we use. I leave it because it is future proof - if anything outside from the definitions in other methods gets allowed in Base for indexing we will allow it.

But I can drop it as it is not strictly needed for now.

Reply:
We don't use that approach elsewhere, do we? I'd rather be consistent. Also the error thrown will be probably be unclear for users.